Some jobs are forever in 'Running' State with WorkflowTaskFailed Error

Some of the workflows remain in Running state while WorkFlowTask failed activity is triggered with the following error. This seems to happen only when I trigger multiple (20-100) instances of the workflow at the same time so it is a little sporadic in nature. The same workflow that works normally, hangs in ‘running’ mode with the error below:

## Expected Behavior
All runs finish with 'complete' status 

## Actual Behavior
Some runs fail with the following bug

```
15:25:20.731 [Workflow Executor taskQueue="Classify", namespace="default": 1568] ERROR i.t.internal.worker.PollerOptions - uncaught exception
java.lang.RuntimeException: Failure processing workflow task. WorkflowId=438cb389-3973-49cf-b298-64cee904ad44, RunId=6810e3eb-be56-4e3b-b810-3c12bf44f76e, Attempt=798
	at io.temporal.internal.worker.WorkflowWorker$TaskHandlerImpl.wrapFailure(WorkflowWorker.java:349)
	at io.temporal.internal.worker.WorkflowWorker$TaskHandlerImpl.wrapFailure(WorkflowWorker.java:279)
	at io.temporal.internal.worker.PollTaskExecutor.lambda$process$0(PollTaskExecutor.java:79)
at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: io.temporal.internal.replay.InternalWorkflowTaskException: Failure handling event 302 of 'EVENT_TYPE_WORKFLOW_TASK_STARTED' type. IsReplaying=false, PreviousStartedEventId=302, workflowTaskStartedEventId=302, Currently Processing StartedEventId=302
	at io.temporal.internal.statemachines.WorkflowStateMachines.handleEvent(WorkflowStateMachines.java:188)
	at io.temporal.internal.replay.ReplayWorkflowRunTaskHandler.handleEvent(ReplayWorkflowRunTaskHandler.java:140)
	at io.temporal.internal.replay.ReplayWorkflowRunTaskHandler.handleWorkflowTaskImpl(ReplayWorkflowRunTaskHandler.java:180)
	at io.temporal.internal.replay.ReplayWorkflowRunTaskHandler.handleWorkflowTask(ReplayWorkflowRunTaskHandler.java:150)
	at io.temporal.internal.replay.ReplayWorkflowTaskHandler.handleWorkflowTaskWithEmbeddedQuery(ReplayWorkflowTaskHandler.java:201)
	at io.temporal.internal.replay.ReplayWorkflowTaskHandler.handleWorkflowTask(ReplayWorkflowTaskHandler.java:114)
	at io.temporal.internal.worker.WorkflowWorker$TaskHandlerImpl.handle(WorkflowWorker.java:319)
	at io.temporal.internal.worker.WorkflowWorker$TaskHandlerImpl.handle(WorkflowWorker.java:279)
	at io.temporal.internal.worker.PollTaskExecutor.lambda$process$0(PollTaskExecutor.java:73)
	... 3 common frames omitted
Caused by: java.lang.RuntimeException: WorkflowTask: failure executing SCHEDULED->WORKFLOW_TASK_STARTED, transition history is [CREATED->WORKFLOW_TASK_SCHEDULED]
	at io.temporal.internal.statemachines.StateMachine.executeTransition(StateMachine.java:140)
	at io.temporal.internal.statemachines.StateMachine.handleHistoryEvent(StateMachine.java:91)
	at io.temporal.internal.statemachines.EntityStateMachineBase.handleEvent(EntityStateMachineBase.java:63)
	at io.temporal.internal.statemachines.WorkflowStateMachines.handleEventImpl(WorkflowStateMachines.java:205)
	at io.temporal.internal.statemachines.WorkflowStateMachines.handleEvent(WorkflowStateMachines.java:173)
	... 11 common frames omitted
Caused by: java.lang.ClassCastException: class scala.runtime.NonLocalReturnControl cannot be cast to class java.lang.Exception (scala.runtime.NonLocalReturnControl is in unnamed module of loader 'app'; java.lang.Exception is in module java.base of loader 'bootstrap')
	at io.temporal.serviceclient.CheckedExceptionWrapper.wrap(CheckedExceptionWrapper.java:56)
	at io.temporal.internal.sync.WorkflowInternal.wrap(WorkflowInternal.java:412)
	at io.temporal.internal.sync.DeterministicRunnerImpl.runUntilAllBlocked(DeterministicRunnerImpl.java:211)
	at io.temporal.internal.sync.SyncWorkflow.eventLoop(SyncWorkflow.java:148)
	at io.temporal.internal.replay.ReplayWorkflowExecutor.eventLoop(ReplayWorkflowExecutor.java:74)
	at io.temporal.internal.replay.ReplayWorkflowRunTaskHandler$EntityManagerListenerImpl.eventLoop(ReplayWorkflowRunTaskHandler.java:329)
	at io.temporal.internal.statemachines.WorkflowStateMachines.eventLoop(WorkflowStateMachines.java:400)
	at io.temporal.internal.statemachines.WorkflowStateMachines.access$500(WorkflowStateMachines.java:71)
	at io.temporal.internal.statemachines.WorkflowStateMachines$WorkflowTaskCommandsListener.workflowTaskStarted(WorkflowStateMachines.java:820)
	at io.temporal.internal.statemachines.WorkflowTaskStateMachine.handleCompleted(WorkflowTaskStateMachine.java:121)
	at io.temporal.internal.statemachines.WorkflowTaskStateMachine.handleStarted(WorkflowTaskStateMachine.java:113)
	at io.temporal.internal.statemachines.FixedTransitionAction.apply(FixedTransitionAction.java:45)
	at io.temporal.internal.statemachines.StateMachine.executeTransition(StateMachine.java:137)
	... 15 common frames omitted
```

## Steps to Reproduce the Problem

  1. kick-off 20-100 jobs of a custom workflow
  2. Most of them would complete successfully while few (1-9) would be in forever running state with workflowTaskFailed Activity
  

## Specifications

  - Version: Temporal Client SDK version 1.3.1
  - Platform: IntelliJ

Hello @Spikhalskiy 

We run a classification workflow that has multiple activities that perform fetching the data, disk operations, and finally interacting with various APIs to get back the results.

Maxim pointed out there might be an edge case unhandled in temporal Java SDK particularly NonLocalReturnControl exception extending Throwable directly in Scala which is not ideal. So was hoping this issue could be fixed.

```
Caused by: java.lang.ClassCastException: class scala.runtime.NonLocalReturnControl cannot be cast to class java.lang.Exception (scala.runtime.NonLocalReturnControl is in unnamed module of loader 'app'; java.lang.Exception is in module java.base of loader 'bootstrap')
	at io.temporal.serviceclient.CheckedExceptionWrapper.wrap(CheckedExceptionWrapper.java:56)
	at io.temporal.internal.sync.WorkflowInternal.wrap(WorkflowInternal.java:412)
```

--- a/temporal-sdk/src/test/java/io/temporal/internal/sync/CheckedExceptionWrapperTest.java
+++ b/temporal-sdk/src/test/java/io/temporal/internal/sync/CheckedExceptionWrapperTest.java
@@ -19,6 +19,8 @@
 
 package io.temporal.internal.sync;
 
+import static org.junit.Assert.assertTrue;
+
 import io.temporal.serviceclient.CheckedExceptionWrapper;
 import io.temporal.workflow.Workflow;
 import org.junit.Assert;
@@ -57,4 +59,13 @@ public class CheckedExceptionWrapperTest {
     Throwable eu = CheckedExceptionWrapper.unwrap(e);
     Assert.assertEquals(e, eu);
   }
+
+  @Test
+  public void customThrowable() {
+    RuntimeException wrapped = CheckedExceptionWrapper.wrap(new CustomThrowable());
+    Throwable unwrapped = CheckedExceptionWrapper.unwrap(wrapped);
+    assertTrue(unwrapped instanceof CustomThrowable);
+  }
+
+  private static class CustomThrowable extends Throwable {}
 }

--- a/temporal-sdk/src/test/java/io/temporal/workflow/failure/WorkflowFailureNonRetryableFlagTest.java
+++ b/temporal-sdk/src/test/java/io/temporal/workflow/failure/WorkflowFailureNonRetryableFlagTest.java
@@ -17,7 +17,7 @@
  *  permissions and limitations under the License.
  */
 
-package io.temporal.workflow;
+package io.temporal.workflow.failure;
 
 import io.temporal.client.WorkflowException;
 import io.temporal.common.RetryOptions;
@@ -45,7 +45,7 @@ public class WorkflowFailureNonRetryableFlagTest {
       SDKTestWorkflowRule.newBuilder().setWorkflowTypes(TestWorkflowNonRetryableFlag.class).build();
 
   @Test
-  public void testWorkflowFailureNonRetryableFlag() {
+  public void nonRetryableFlag() {
     RetryOptions workflowRetryOptions =
         RetryOptions.newBuilder()
             .setInitialInterval(Duration.ofSeconds(1))
@@ -81,11 +81,7 @@ public class WorkflowFailureNonRetryableFlagTest {
 
     @Override
     public String execute(String testName) {
-      AtomicInteger count = retryCount.get(testName);
-      if (count == null) {
-        count = new AtomicInteger();
-        retryCount.put(testName, count);
-      }
+      AtomicInteger count = retryCount.computeIfAbsent(testName, ignore -> new AtomicInteger());
       int c = count.incrementAndGet();
       ApplicationFailure f =
           ApplicationFailure.newFailure("simulated " + c, "foo", "details1", 123);

--- a/temporal-sdk/src/test/java/io/temporal/workflow/failure/WorkflowFailureNonStandardThrowableTest.java
+++ b/temporal-sdk/src/test/java/io/temporal/workflow/failure/WorkflowFailureNonStandardThrowableTest.java
@@ -1,0 +1,128 @@
+/*
+ *  Copyright (C) 2020 Temporal Technologies, Inc. All Rights Reserved.
+ *
+ *  Copyright 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *  Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"). You may not
+ *  use this file except in compliance with the License. A copy of the License is
+ *  located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ *  or in the "license" file accompanying this file. This file is distributed on
+ *  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *  express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+
+package io.temporal.workflow.failure;
+
+import static org.junit.Assert.*;
+
+import io.temporal.client.WorkflowException;
+import io.temporal.common.RetryOptions;
+import io.temporal.failure.ApplicationFailure;
+import io.temporal.testing.internal.SDKTestOptions;
+import io.temporal.testing.internal.SDKTestWorkflowRule;
+import io.temporal.worker.WorkflowImplementationOptions;
+import io.temporal.workflow.shared.TestWorkflows.TestWorkflow1;
+import java.time.Duration;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TestName;
+
+/** https://github.com/temporalio/sdk-java/issues/744 */
+public class WorkflowFailureNonStandardThrowableTest {
+
+  public static class NonStandardThrowable extends Throwable {}
+
+  private static final Map<String, AtomicInteger> retryCount = new ConcurrentHashMap<>();
+
+  @Rule public TestName testName = new TestName();
+
+  @Rule
+  public SDKTestWorkflowRule testWorkflowRule =
+      SDKTestWorkflowRule.newBuilder()
+          .setWorkflowTypes(
+              WorkflowImplementationOptions.newBuilder()
+                  .setFailWorkflowExceptionTypes(NonStandardThrowable.class)
+                  .build(),
+              TestWorkflowNonStandardThrowable.class)
+          .build();
+
+  @Test
+  public void nonStandardThrowable() {
+    TestWorkflow1 workflowStub =
+        testWorkflowRule
+            .getWorkflowClient()
+            .newWorkflowStub(
+                TestWorkflow1.class,
+                SDKTestOptions.newWorkflowOptionsWithTimeouts(testWorkflowRule.getTaskQueue()));
+
+    try {
+      workflowStub.execute(testName.getMethodName());
+      fail();
+    } catch (WorkflowException e) {
+      assertTrue(e.getCause() instanceof ApplicationFailure);
+      ApplicationFailure applicationFailure = (ApplicationFailure) e.getCause();
+      assertEquals(NonStandardThrowable.class.getName(), applicationFailure.getType());
+    }
+  }
+
+  @Test
+  public void nonStandardThrowableSuccessOnSecondAttempt() {
+    RetryOptions workflowRetryOptions =
+        RetryOptions.newBuilder()
+            .setInitialInterval(Duration.ofMillis(1))
+            .setMaximumAttempts(2)
+            .setBackoffCoefficient(1.0)
+            .build();
+    TestWorkflow1 workflowStub =
+        testWorkflowRule
+            .getWorkflowClient()
+            .newWorkflowStub(
+                TestWorkflow1.class,
+                SDKTestOptions.newWorkflowOptionsWithTimeouts(testWorkflowRule.getTaskQueue())
+                    .toBuilder()
+                    .setRetryOptions(workflowRetryOptions)
+                    .build());
+
+    String result = workflowStub.execute(testName.getMethodName());
+    assertEquals("success", result);
+    assertEquals(
+        "Success is expected on a second run only",
+        2,
+        retryCount.get(testName.getMethodName()).get());
+  }
+
+  public static class TestWorkflowNonStandardThrowable implements TestWorkflow1 {
+
+    @Override
+    public String execute(String testName) {
+      AtomicInteger count = retryCount.computeIfAbsent(testName, ignore -> new AtomicInteger());
+      int c = count.incrementAndGet();
+      if (c <= 1) {
+        rethrow(new NonStandardThrowable());
+        // unreachable
+        return "fail";
+      } else {
+        return "success";
+      }
+    }
+  }
+
+  private static <T extends Throwable> void rethrow(Throwable e) throws T {
+    if (e instanceof RuntimeException) {
+      throw (RuntimeException) e;
+    } else {
+      @SuppressWarnings("unchecked")
+      T toRethrow = (T) e;
+      throw toRethrow;
+    }
+  }
+}

--- a/temporal-serviceclient/src/main/java/io/temporal/serviceclient/CheckedExceptionWrapper.java
+++ b/temporal-serviceclient/src/main/java/io/temporal/serviceclient/CheckedExceptionWrapper.java
@@ -53,7 +53,7 @@ public final class CheckedExceptionWrapper extends RuntimeException {
     if (e instanceof RuntimeException) {
       return (RuntimeException) e;
     }
-    return new CheckedExceptionWrapper((Exception) e);
+    return new CheckedExceptionWrapper(e);
   }
 
   /**
@@ -64,7 +64,7 @@ public final class CheckedExceptionWrapper extends RuntimeException {
     return e instanceof CheckedExceptionWrapper ? e.getCause() : e;
   }
 
-  private CheckedExceptionWrapper(Exception e) {
+  private CheckedExceptionWrapper(Throwable e) {
     super(e);
   }
 }

